clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,202 @@
+}
+
+func updateFailureMetrics(apiAvailabilityMetrics *apiAvailabilityMetrics) {
+	// increment total number of failed probes count, current number of failed probes by 1
+	apiAvailabilityMetrics.numFailures++
+	apiAvailabilityMetrics.consecutiveFailedProbes++
+	// if the previous probe was not a failure, then set the previous probe failure flag to true.
+	// This would be used by next probe
+	if !apiAvailabilityMetrics.prevProbeFailure {
+		apiAvailabilityMetrics.prevProbeFailure = true
+	}
+}
+
+func updateSuccessMetrics(apiAvailabilityMetrics *apiAvailabilityMetrics) {
+	// if the probe succeeded, then increase the number of successful probes count by 1
+	apiAvailabilityMetrics.numSuccess++
+	// if this successful probe was received after failing probe(s)
+	if apiAvailabilityMetrics.prevProbeFailure {
+		// Check if the recently concluded consecutive failed number of probes is
+		// higher than the max consecutive failed number of probes so far
+		// if yes, then Update max consecutive failed probes
+		if apiAvailabilityMetrics.consecutiveFailedProbes > apiAvailabilityMetrics.maxConsecutiveFailedProbes {
+			apiAvailabilityMetrics.maxConsecutiveFailedProbes = apiAvailabilityMetrics.consecutiveFailedProbes
+		}
+		// this probe won't be a failure for the next probe
+		apiAvailabilityMetrics.prevProbeFailure = false
+		// reset current consecutive failed number of probes
+		apiAvailabilityMetrics.consecutiveFailedProbes = 0
+	}
+}
+
+// Execute starts the api-server healthz probe end point from start action and
+// collect metrics in gather.
+func (a *apiAvailabilityMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+
+	SSHToMasterSupported := config.ClusterFramework.GetClusterConfig().SSHToMasterSupported
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	probeDuration, err := util.GetInt(config.Params, "duration")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if a.isRunning {
+			klog.Infof("%s: measurement already running", a)
+			return nil, nil
+		}
+		return a.start(config, SSHToMasterSupported, probeDuration)
+	case "gather":
+		a.stop()
+		availabilityPercentage := (a.apiAvailabilityMetrics.numSuccess / a.apiAvailabilityMetrics.numSuccess) * 100
+		longestUnavailableDuration := a.apiAvailabilityMetrics.maxConsecutiveFailedProbes * probeDuration
+		availabilityMetrics := &apiAvailabilityOutput{AvailabilityPercentage: availabilityPercentage, LongestUnavailableDuration: longestUnavailableDuration}
+		content, err := util.PrettyPrintJSON(availabilityMetrics)
+		if err != nil {
+			return nil, err
+		}
+		summary := measurement.CreateSummary(etcdMetricsMetricName, "json", content)
+		return []measurement.Summary{summary}, nil
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (a *apiAvailabilityMeasurement) stop() {
+	if !a.isRunning {
+		return
+	}
+	close(a.stopCh)
+	a.wg.Wait()
```

Comment:
That doesn't fully solve the problem - if the last period was the longest one, it won't be reported.

Reply:
Can you plz help me understand this scenario? Are you saying that when gather was called, there could have been another probe which would have made it the longest failing one and because I am closing the channel, it will miss that probe?
I have limited understanding of go routines but if the above scenario is correct, then does it mean that the channel stopping would override any currently executing probe? Should I wait before closing to check if any probes are ongoing?

Reply:
This is the problem:
https://github.com/kubernetes/perf-tests/pull/1162/files#diff-769e85d1bdfa456cbbe72288f2ba2ab5R140-R142

If you have gazillions failure in a row at the end of the test, you will not include it in your summary.

Reply:
I added the comparision during the gather flow as well.